rest-server/src/config/job.js

```diff
@@ -74,15 +74,20 @@ const jobConfigSchema = Joi.object().keys({
           .default([]),
         command: Joi.string()
           .required(),
+        minFailedTaskCount: Joi.number()
+          .integer()
+          .min(1)
+          .optional(),
+        minSucceededTaskCount: Joi.number()
+          .integer()
+          .min(1)
+          .optional(),
```

Comment:
To distinguish field NOT EXIST and field is set to NULL (To allow use can set minFailedTaskCount = null).

Can we make:
  minFailedTaskCount.default(1)
  minSucceededTaskCount.default(null)

And 
        'applicationCompletionPolicy': {
          'minFailedTaskCount': data.taskRoles[i].minFailedTaskCount,
          'minSucceededTaskCount': data.taskRoles[i].minSucceededTaskCount,
        },

Reply:
Default value means if user doesn't set the field then the field will be set to default value, which is the same as current code.
We can make `minFailedTaskCount.default(1)` but can't make `minSucceededTaskCount.default(null)` because `minSucceededTaskCount` is limited to integer in schema.

Reply:
If user set the minFailedTaskCount = null, from below line:

'minFailedTaskCount': data.taskRoles[i].minFailedTaskCount || 1,

What will you send to launcher? null or 1?
I expect you also send null to launcher.




---
In reply to: [202507760](https://github.com/Microsoft/pai/pull/879#discussion_r202507760) [](ancestors = 202507760)

Reply:
Because the minFailedTaskCount is limited to integer, user can not set it to null. I suggest we can use 0 to replace null in config file.
And what's difference between minFailedTaskCount=null and minFailedTaskCount=taskNumber?

Reply:
Do we have any way to let integer support null?
Such as value is null, undefined or NaN.


[https://stackoverflow.com/questions/16474796/set-javascript-variable-null-or-leave-undefined](https://stackoverflow.com/questions/16474796/set-javascript-variable-null-or-leave-undefined)
"undefined" means this value does not exist. typeof returns "undefined"

"null" means this value exists with an empty value. When you use typeof to test for "null", you will see that it's an object. Other case when you serialize "null" value to backend server like asp.net mvc, the server will receive "null", but when you serialize "undefined", the server is unlikely to receive a value.

---
In reply to: [203658064](https://github.com/Microsoft/pai/pull/879#discussion_r203658064) [](ancestors = 203658064,203646504)

Reply:
Can we use 0 to support null? Because minFailedTaskCount is greater or equal than 1 at present.

Reply:
Please use -1 to substitute null if js really cannot accept null as int value.

---
In reply to: [203690975](https://github.com/Microsoft/pai/pull/879#discussion_r203690975) [](ancestors = 203690975)

Reply:
updated, allow `null` in `minFailedTaskCount` and `minSucceededTaskCount`